glommio/src/executor/stall.rs

```diff
@@ -0,0 +1,543 @@
+
+    #[allow(dead_code)]
+    #[derive(Debug)]
+    pub struct TestStallDetection {
+        executor: usize,
+        queue_handle: TaskQueueHandle,
+        queue_name: String,
+        trace: backtrace::Backtrace,
+        budget: Duration,
+        overage: Duration,
+    }
+
+    #[derive(Debug)]
+    struct InnerTestHandler {
+        detections: Vec<TestStallDetection>,
+    }
+
+    #[derive(Clone, Debug)]
+    struct TestHandler {
+        inner: Arc<RwLock<InnerTestHandler>>,
+        signal: u8,
+    }
+
+    impl TestHandler {
+        fn new(signal: u8) -> Self {
+            TestHandler {
+                inner: Arc::new(RwLock::new(InnerTestHandler {
+                    detections: Vec::new(),
+                })),
+                signal,
+            }
+        }
+    }
+
+    impl StallDetectionHandler for TestHandler {
+        fn high_water_mark(
+            &self,
+            _queue_handle: TaskQueueHandle,
+            max_expected_runtime: Duration,
+        ) -> Option<Duration> {
+            Some(
+                Duration::from_millis((max_expected_runtime.as_millis() as f64 * 0.1) as u64)
+                    .max(Duration::from_millis(10)),
+            )
+        }
+
+        fn signal(&self) -> u8 {
+            self.signal
+        }
+
+        fn stall(&self, detection: StallDetection<'_>) {
+            let mut inner = self.inner.write().unwrap();
+            inner.detections.push(TestStallDetection {
+                executor: detection.executor,
+                queue_handle: detection.queue_handle,
+                queue_name: detection.queue_name.to_owned(),
+                trace: detection.trace,
+                budget: detection.budget,
+                overage: detection.overage,
+            });
+        }
+    }
+
+    #[test]
+    fn executor_stall_detector() {
+        let stall_handler = TestHandler::new(nix::libc::SIGUSR1 as u8);
+        LocalExecutorBuilder::default()
+            .detect_stalls(Some(Box::new(stall_handler.clone())))
+            .preempt_timer(Duration::from_millis(50))
+            .make()
+            .unwrap()
+            .run(async {
+                // will trigger the stall detector because we go over budget
+                thread::sleep(Duration::from_millis(100));
+
+                let exec = crate::executor();
+                assert!(stall_handler.inner.read().unwrap().detections.is_empty());
+
+                exec.yield_task_queue_now().await; // yield the queue
+
+                assert!(
+                    stall_handler
+                        .inner
+                        .write()
+                        .unwrap()
+                        .detections
+                        .pop()
+                        .is_some()
+                );
+
+                // no stall because < 50ms of un-cooperativeness
+                thread::sleep(Duration::from_millis(40));
+
+                assert!(stall_handler.inner.read().unwrap().detections.is_empty());
```

Comment:
This will never trigger. Not because you have not stalled but because the handler is called after the task queue has returned. Since you are executing in the queue, it will always be empty.

You need to move this assertion after the yielding point just below.